File: frontend/src/api/adminOps.ts
import { api } from "./client";
import type { ChallengeState } from "./challenges";

export interface Adjustment {
  id: string;
  user_id: string | null;
  team_id: string | null;
  subject_name: string | null;
  points: number;
  reason: string;
  created_by_user_id: string | null;
  created_by_name: string | null;
  reverses_id: string | null;
  /** Set when a later entry cancelled this one. Shown struck through, never hidden. */
  reversed_by_id: string | null;
  created_at: string;
}

export type ReportStatus = "open" | "acknowledged" | "dismissed" | "resolved";

export interface Report {
  id: string;
  challenge_id: string;
  challenge_title: string | null;
  user_id: string;
  reporter_name: string | null;
  message: string;
  status: ReportStatus;
  resolution_note: string | null;
  created_at: string;
}

export interface ChallengeHealth {
  challenge_id: string;
  title: string;
  state: ChallengeState;
  solve_count: number;
  attempt_count: number;
  open_reports: number;
  suspected_broken: boolean;
  suspiciously_easy: boolean;
}

export interface Dashboard {
  generated_at: string;
  event: {
    name: string | null;
    starts_at: string | null;
    ends_at: string | null;
    running: boolean;
    server_time: string;
  };
  pulse: {
    solves_5m: number;
    solves_15m: number;
    solves_60m: number;
    submissions_15m: number;
    active_players_15m: number;
  };
  attention: {
    open_reports: number;
    pending_approvals: number;
    drafts_after_start: number;
    published_without_answers: { challenge_id: string; title: string }[];
    suspected_broken: { challenge_id: string; title: string; attempts: number }[];
  };
  containers: { available: boolean; note: string };
}

export interface AuditEntry {
  id: string;
  action: string;
  actor_user_id: string | null;
  actor_name: string | null;
  target_type: string;
  target_id: string | null;
  reason: string | null;
  metadata: Record<string, unknown>;
  request_id: string | null;
  created_at: string;
}

export const getDashboard = () => api.get<Dashboard>("/admin/dashboard");
export const getChallengeHealth = () => api.get<ChallengeHealth[]>("/admin/challenge-health");
export const listAdjustments = () => api.get<Adjustment[]>("/admin/adjustments");

export const createAdjustment = (input: {
  user_id?: string;
  team_id?: string;
  points: number;
  reason: string;
}) => api.post<Adjustment>("/admin/adjustments", input);

export const reverseAdjustment = (id: string, reason: string) =>
  api.post<Adjustment>(`/admin/adjustments/${id}/reverse`, { reason });

export const listReports = (status?: ReportStatus) =>
  api.get<Report[]>(`/admin/reports${status ? `?status=${status}` : ""}`);

export const triageReport = (id: string, status: ReportStatus, note?: string) =>
  api.post<{ message: string }>(`/admin/reports/${id}/status`, { status, note: note ?? null });

export const listAudit = (action?: string) =>
  api.get<AuditEntry[]>(`/admin/audit-log${action ? `?action=${action}` : ""}`);

export const reportChallenge = (challengeId: string, message: string) =>
  api.post<Report>(`/challenges/${challengeId}/report`, { message });

File: backend/app/schemas/admin_ops.py
"""Request and response models for operational admin actions."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.models.challenge import ChallengeState, PreReleaseState
from app.models.report import ReportStatus


class CreateAdjustmentRequest(BaseModel):
    #: Exactly one target. A party adjustment belongs to the party itself, not
    #: to its members and not to its leader.
    user_id: UUID | None = None
    team_id: UUID | None = None
    points: int = Field(ge=-100_000, le=100_000)
    #: Required. An unexplained score change is indefensible after the event.
    reason: str = Field(min_length=3, max_length=500)

    @model_validator(mode="after")
    def _exactly_one_target(self) -> "CreateAdjustmentRequest":
        if (self.user_id is None) == (self.team_id is None):
            raise ValueError("Provide exactly one of user_id or team_id.")
        return self


class ReverseAdjustmentRequest(BaseModel):
    reason: str = Field(min_length=3, max_length=500)


class AdjustmentResponse(BaseModel):
    id: UUID
    user_id: UUID | None
    team_id: UUID | None
    subject_name: str | None
    points: int
    reason: str
    created_by_user_id: UUID | None
    created_by_name: str | None
    reverses_id: UUID | None
    #: Set when a later entry cancelled this one. Shown struck through, never hidden.
    reversed_by_id: UUID | None
    created_at: datetime


class ReportRequest(BaseModel):
    message: str = Field(min_length=5, max_length=1000)


class ReportResponse(BaseModel):
    id: UUID
    challenge_id: UUID
    challenge_title: str | None
    user_id: UUID
    reporter_name: str | None
    message: str
    status: ReportStatus
    resolution_note: str | None
    created_at: datetime


class TriageReportRequest(BaseModel):
    status: ReportStatus
    note: str | None = Field(default=None, max_length=500)


class ChallengeHealthResponse(BaseModel):
    challenge_id: UUID
    title: str
    state: ChallengeState
    solve_count: int
    attempt_count: int
    open_reports: int
    suspected_broken: bool
    suspiciously_easy: bool


class BulkReleaseRequest(BaseModel):
    challenge_ids: list[UUID] = Field(min_length=1, max_length=200)
    #: Null clears the schedule, making the challenge follow its own state.
    release_at: datetime | None = None
    pre_release_state: PreReleaseState | None = None


class AuditEntryResponse(BaseModel):
    id: UUID
    action: str
    actor_user_id: UUID | None
    actor_name: str | None
    target_type: str
    target_id: UUID | None
    reason: str | None
    metadata: dict[str, Any]
    request_id: str | None
    created_at: datetime


class AuditPageResponse(BaseModel):
    """One page of the log, with the size of the whole result set.

    Paginated where spec 041 deliberately refused to paginate the challenge
    list: that is 242 rows with a known ceiling, while the log grows unbounded
    across five days of approvals, kicks and edits, and has no natural grouping
    to collapse under.
    """

    total: int
    entries: list[AuditEntryResponse]


class PlayDataGroup(BaseModel):
    """What one clearable group currently holds (spec 043)."""

    group: str
    #: Written for an admin, not for a schema — this is the confirm dialog's text.
    label: str
    rows: int


class ResetPlayDataRequest(BaseModel):
    """Groups are named explicitly, never defaulted. The difference between
    clearing one thing and clearing all fourteen must not come down to a
    missing field."""

    groups: list[str] = Field(min_length=1)


class ResetPlayDataResponse(BaseModel):
    #: Table name -> rows removed.
    deleted: dict[str, int]
    total: int
